res/controllers/Pioneer-DDJ-400-script.js

```diff
@@ -0,0 +1,666 @@
+// ****************************************************************************
+//
+//  Implemented (as per manufacturer's manual):
+//      * Mixer Section (Faders, EQ, Filter, Gain, Cue)
+//      * Browsing and loading + Waveform zoom (shift)
+//      * Jogwheels, Scratching, Bending, Loop adjust
+//      * Cycle Temporange
+//      * Beat Sync
+//      * Hot Cue Mode
+//      * Beat Loop Mode
+//      * Beat Jump Mode
+//      * Sampler Mode
+//
+//  Custom (Mixxx specific mappings):
+//      * BeatFX: Assigned Effect Unit 1
+//                < LEFT focus EFFECT1
+//                > RIGHT focus EFFECT2
+//                v FX_SELECT focus EFFECT3.
+//                ON/OFF toggles focused effect slot
+//                SHIFT + ON/OFF disables all three effect slots.
+//                SHIFT + < loads previous effect
+//                SHIFT + > loads next effect
+//
+//      * 32 beat jump forward & back (Shift + </> CUE/LOOP CALL arrows)
+//      * Toggle quantize (Shift + channel cue)
+//
+//  Not implemented (after discussion and trial attempts):
+//      * Loop Section:
+//        * -4BEAT auto loop (hacky---prefer a clean way to set a 4 beat loop
+//                            from a previous position on long press)
+//
+//        * CUE/LOOP CALL - memory & delete (complex and not useful. Hot cues are sufficient)
+//
+//      * Secondary pad modes (trial attempts complex and too experimental)
+//        * Keyboard mode
+//        * Pad FX1
+//        * Pad FX2
+//        * Keyshift mode
+
+var PioneerDDJ400 = {};
+
+PioneerDDJ400.lights = {
+    beatFx: {
+        status: 0x94,
+        data1: 0x47,
+    },
+    deck1: {
+        vuMeter: {
+            status: 0xB0,
+            data1: 0x02,
+        },
+        playPause: {
+            status: 0x90,
+            data1: 0x0B,
+        },
+        shiftPlayPause: {
+            status: 0x90,
+            data1: 0x47,
+        },
+        cue: {
+            status: 0x90,
+            data1: 0x0C,
+        },
+        shiftCue: {
+            status: 0x90,
+            data1: 0x48,
+        },
+    },
+    deck2: {
+        vuMeter: {
+            status: 0xB0,
+            data1: 0x02,
+        },
+        playPause: {
+            status: 0x91,
+            data1: 0x0B,
+        },
+        shiftPlayPause: {
+            status: 0x91,
+            data1: 0x47,
+        },
+        cue: {
+            status: 0x91,
+            data1: 0x0C,
+        },
+        shiftCue: {
+            status: 0x91,
+            data1: 0x48,
+        },
+    },
+};
+
+// Store timer IDs
+PioneerDDJ400.timers = {};
+
+// Jog wheel constants
+PioneerDDJ400.vinylMode = true;
+PioneerDDJ400.alpha = 1.0/8;
+PioneerDDJ400.beta = PioneerDDJ400.alpha/32;
+
+// Multiplier for fast seek through track using SHIFT+JOGWHEEL
+PioneerDDJ400.fastSeekScale = 150;
+PioneerDDJ400.bendScale = 0.9;
+
+PioneerDDJ400.tempoRanges = [0.06, 0.10, 0.16, 0.25]; // WIDE = 25%?
```

Comment:
WIDE is 100% in rekordbox, but this seems a bit overkill for the DDJ-400. 25% is fine in my opinion.